draft-irtf-cfrg-vdaf.md

```diff
@@ -342,70 +359,284 @@ measurement process is as follows:
+> https://github.com/ietf-wg-ppm/ppm-specification/issues/19.
+
+## Execution of a DAF {#daf-execution}
+
+Securely executing a DAF involves emulating the following procedure.
+
+~~~
+def run_daf(Daf,
+            agg_param: Daf.AggParam,
+            measurements: Vec[Daf.Measurement]):
+    out_shares = []
+    for measurement in measurements:
+        # Each Client shards its measurement into input shares and
+        # distributes them among the Aggregators.
+        input_shares = Daf.measurement_to_input_shares(measurement)
+
+        # Each Aggregator prepares its input share for aggregation.
+        outbound = []
+        for j in range(Daf.SHARES):
+            outbound.append(Daf.prep(agg_param, input_shares[j]))
+        out_shares.append(outbound)
+
+    # Each Aggregator aggregates its output shares into an aggregate
+    # share and it to the Collector.
+    agg_shares = []
+    for j in range(Daf.SHARES):
+        out_shares_j = [out[j] for out in out_shares]
+        agg_share_j = Daf.out_shares_to_agg_share(agg_param,
+                                                  out_shares_j)
+        agg_shares.append(agg_share_j)
+
+    # Collector unshards the aggregate result.
+    agg_result = Daf.agg_shares_to_result(agg_param, agg_shares)
+    return agg_result
+~~~
+{: #run-daf title="Execution of a DAF."}
+
+The inputs to this procedure are the same as the aggregation function computed by
+the DAF: An aggregation parameter and a sequence of measurements. The procedure
+prescribes how a DAF is executed in a "benign" environment in which there is no
+adversary and the messages are passed among the protocol participants over
+secure point-to-point channels. In reality, these channels need to be
+instantiated by some "wrapper protocol", such as {{I-D.draft-gpew-priv-ppm}},
+that realizes these channels using suitable cryptographic mechanisms. Moreover,
+some fraction of the Aggregators (or Clients) may be malicious and diverge from
+their prescribed behaviors. {{security}} describes the execution of the DAF in
+various adversarial environments and what properties the wrapper protocol needs
+to provide in each.
+
 # Definition of VDAFs {#vdaf}
 
-A concrete VDAF specifies the algorithms involved in evaluating an aggregation
-function across a batch of inputs. This section specifies the interfaces of
-these algorithms as they would be exposed to applications.
+Like DAFs described in the previous section, a VDAF scheme is used to compute a
+particular aggregation function over a set of Client-generated measurements.
+Evaluation of a VDAF involves the same four stages as for DAFs: Sharding,
+Preparation, Aggregation, and Unsharding. However, the Preparation stage will
+require interaction among the Aggregators in order to facilitate verifiability
+of the computation's correctness. Accommodating this interaction will require
+syntactic changes, as well as a new Setup phase for distributing long-lived
+verification parameters.
```

Comment:
> However, the Preparation stage will require interaction among the Aggregators in order to facilitate verifiability of the computation's correctness.

I think it makes sense to split up the preparation (as in, turning an input share into an aggregation share), and the verification.
The reason is that the outer protocol using a VDAF might want to "opportunistically" aggregate shares, in particular for Poplar, where there are many rounds of Preparation, Aggregation, and Unsharding. This would save communication rounds, since the verification on one layer (of the Poplar tree) can be done in parallel with computing the aggregation shares on the next layer. In case a share turns out to be invalid, the corresponding aggregation shares can just be subtracted from the aggregated result.

Reply:
To be clear:
* Prepare takes input shares to output shares; and
* Aggregation takes output shares to aggregate shares.

> [...] the verification on one layer (of the Poplar tree) can be done in parallel with computing the aggregation shares on the next layer. In case a share turns out to be invalid, the corresponding aggregation shares can just be subtracted from the aggregated result.

Very Interesting idea. The current text points to two reasons for not breaking Prepare abstraction this way:
1. Not every VDAF is compatible with this change. E.g., in Prio+ the output shares are only computed after multiple rounds of interaction.
2. Releasing unverified output shares early may lead to them being misused.

Thus, a change here would be significant and would need to be thought through. I'd suggest we discuss it in a new issue.

Reply:
> Not every VDAF is compatible with this change. E.g., in Prio+ the output shares are only computed after multiple rounds of interaction.
I'm not asking for every VDAF to support this opportunistic aggregation, but instead to *allow* it in places where it's supported by both the VDAF and the outer aggregation protocol.

I opened https://github.com/cfrg/draft-irtf-cfrg-vdaf/issues/61 to continue this discussion.